packages/braintree-integration/src/braintree-paypal-ach/braintree-paypal-ach-payment-strategy.ts

```diff
@@ -69,7 +74,103 @@ export default class BraintreePaypalAchPaymentStrategy implements PaymentStrateg
+
+            await this.paymentIntegrationService.submitOrder(order, options);
+            await this.paymentIntegrationService.submitPayment(submitPaymentPayload);
+        } catch (error) {
+            this.handleError(error);
+        }
+    }
+
+    finalize(): Promise<void> {
+        return Promise.reject(new OrderFinalizationNotRequiredError());
+    }
+
+    async deinitialize(): Promise<void> {
+        this.getMandateText = undefined;
+
+        return Promise.resolve();
+    }
+
+    private async preparePaymentData(
+        nonce: string | null,
+        payment: OrderPaymentRequestBody,
+    ): Promise<Payment> {
+        const { paymentData = {} } = payment;
+
+        if (!nonce) {
+            if (!isVaultedInstrument(paymentData)) {
+                throw new PaymentArgumentInvalidError(['payment.paymentData']);
+            }
+
+            return {
+                methodId: payment.methodId,
+                paymentData: {
+                    instrumentId: paymentData.instrumentId,
+                },
+            };
+        }
+
+        if (!isUsBankAccountInstrumentLike(paymentData)) {
+            throw new PaymentArgumentInvalidError(['payment.paymentData']);
+        }
+
+        const sessionId = await this.braintreeIntegrationService.getSessionId();
+
+        const state = this.paymentIntegrationService.getState();
+
+        const { email } = state.getCheckoutOrThrow().billingAddress || { email: null };
+
+        const { shouldSaveInstrument, shouldSetAsDefaultInstrument, routingNumber, accountNumber } =
+            paymentData;
+
+        const paymentPayload = {
+            formattedPayload: {
+                vault_payment_instrument: shouldSaveInstrument || null,
+                set_as_default_stored_instrument: shouldSetAsDefaultInstrument || null,
+                device_info: sessionId || null,
+                ach_account: {
+                    routing_number: routingNumber,
+                    last4: accountNumber.substr(-4),
+                    token: nonce,
+                    email: email || null,
+                },
+            },
+        };
+
+        return {
+            methodId: payment.methodId,
+            paymentData: paymentPayload,
+        };
+    }
+
+    private async tokenizePayment(payment: OrderPaymentRequestBody) {
+        const { paymentData = {} } = payment;
+
+        if (isVaultedInstrument(paymentData)) {
+            if (!this.paymentMethod) {
+                throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
+            }
+
+            const {
+                config: { isVaultingEnabled },
+            } = this.paymentMethod;
+
+            if (!isVaultingEnabled) {
+                throw new InvalidArgumentError(
+                    'Vaulting is disabled but a vaulted instrument was being used for this transaction',
+                );
+            }
+
+            return null;
+        }
```

Comment:
In this case, to avoid using `this.paymentMethod` we can write code a bit different:
If you check OrderPaymentRequestBody interface, you can see that you have access to payment `methodId` and `gatewayId`. So instead of creating and using `this.paymentMethod` + handle the error all over the place we can get specific paymentMethod from checkout state.

```suggestion
        const {
            methodId,
            paymentData = {},
        } = payment;

        if (isVaultedInstrument(paymentData)) {
            const state = this.paymentIntegrationService.getState();
            const { config } = state.getPaymentMethodOrThrow(methodId);

            if (!config.isVaultingEnabled) {
                throw new InvalidArgumentError(
                    'Vaulting is disabled but a vaulted instrument was being used for this transaction',
                );
            }

            return null;
        }
```